### PR TITLE
Ensure CoreDNS running when Corefile migration doesn't support current version

### LINKED ISSUE #88725: Coredns fails to start if Corefile migration fails
**What happened**:
Coredns failed to start in 1.17 deployment when it tried to retain a backed up Corefile with error message: Caddyfile via flag: open /etc/coredns/Corefile: no such file or directory

**What you expected to happen**:
Coredns starts normally with the backed-up Corefile

**How to reproduce it (as minimally and precisely as possible)**:
Deploy a k8s 1.17 cluster and trigger coredns Corefile migration and make it fail. In our case, this was triggered in a three master -setup with 'old' method of running 'kubeadm init' for each master. Also the core-dns docker image used was tagged with '1.6.7-1', apparently causing the corefile migration version check to fail

**Anything else we need to know?**:
This seems to be caused by
https://github.com/kubernetes/kubernetes/blob/39ed64ec4c3cae32ac1e707a03017f5986efc168/cmd/kubeadm/app/phases/addons/dns/dns.go#L453

Both 'key' and 'path' are set to 'Corefile-backup' when this function is called. Should the 'path' always stay as 'Corefile' because that is where the container expects to find the config file?

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:51:57Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:47:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
Custom deployment in Openstack environment
- OS (e.g: `cat /etc/os-release`):
NAME="SLES"
VERSION="15-SP1"
VERSION_ID="15.1"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP1"
- Kernel (e.g. `uname -a`):
4.12.14-197.29-default
- Install tools:
kubeadm
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by rajansandeep**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR aims to fix a bug in the current released version of kubeadm that had 2 consequences:

1. After a successful corefile migration, the current deployment of CoreDNS was not pointed to the backup copy, such that if the Deployment rollout of the new version of CoreDNS failed, a rollback could fail (if the new config was not compatible with the old version).
2. After a failed corefile migration, the deployment of CoreDNS was left in a failed rollout state. This was issue #88725.
**Which issue(s) this PR fixes**:

Fixes #88725 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 1]
```diff
@@ -371,31 +371,18 @@ func IsCoreDNSConfigMapMigrationRequired(corefile string) bool {
...
-	}
-
 	// Since the current configuration present is not the default version, try and migrate it.
 	updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
 	if err != nil {
 		return errors.Wrap(err, "unable to migrate CoreDNS ConfigMap")
 	}
 
+	// Point the CoreDNS deployment to the `Corefile-backup` data.
+	if err := patchCoreDNSDeployment(client, "Corefile-backup"); err != nil {
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: This step points CoreDNS Deployment to a Configmap value that doesn't exist until the next step (or exists from a prior kubeadm upgrade). IMO, we should create the backup first, then point the deployment to it.

Minor nit: the name of the function `patchCoreDNSDeployment()` seems too generic for what it does.  I think `setCorefile()` or something like that would be better.

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 2]
```diff
@@ -408,6 +390,12 @@ func migrateCoreDNSCorefile(client clientset.Interface, cm *v1.ConfigMap, corefi
...
 		return errors.Wrap(err, "unable to update the CoreDNS ConfigMap")
 	}
+
+	// Point the CoreDNS deployment to the `Corefile-backup` data.
+	if err := setCorefile(client, "Corefile-backup"); err != nil {
+		return err
+	}
+
 	fmt.Println("[addons] Migrating CoreDNS Corefile")
 	changes, err := migration.Deprecated(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile)
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: Should `setCorefile()` be after this too?  Just in case this fails?

**Review by rosti — CHANGES_REQUESTED** [review]
Thanks @rajansandeep !
I have a bit of a problem here. `migrateCoreDNSCorefile` is supposed to be only called whenever `IsCoreDNSConfigMapMigrationRequired` returns true. I do think that the latter function is really supposed to return true and migration to be performed whenever CoreDNS is surely not going to start with the old config (hence a migration is required).
I am not OK to do unnecessary migrations as these can change a Corefile beyond recognition and can cause problems like the one with the failed migration.
If a migration is failing and it's required (CoreDNS can't work without it), then a failure to start is genuine.
With the above in mind, it's probably better to investigate why `IsCoreDNSConfigMapMigrationRequired` is returning true more often than it shouldn't.
/hold

**Comment by chrisohaver** [comment]
`IsCoreDNSConfigMapMigrationRequired ` is perhaps poorly named.  It's function is to determine whether or not we can deploy a default CoreDNS config over top of the existing CoreDNS config.  IOW, if the current CoreDNS config is at default (not modified),  then we don't need to migrate the default, we can just re-deploy the new default over top of it.

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 3]
```diff
@@ -408,6 +421,12 @@ func migrateCoreDNSCorefile(client clientset.Interface, cm *v1.ConfigMap, corefi
 	}, metav1.UpdateOptions{}); err != nil {
 		return errors.Wrap(err, "unable to update the CoreDNS ConfigMap")
 	}
+
+	// Point the CoreDNS deployment to the `Corefile-backup` data.
+	if err := setCorefile(client, "Corefile-backup"); err != nil {
+		return err
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: If this errors, then the original coredns is left pointing at a migrated corefile, which could result in a coredns that wont start.  Is this a concern?  Will this ever error? Fix would be only add `Corefile-backup` in the above configmap update, and then update the `Corefile` value after calling `setCorefile(client, "Corefile-backup")`.  This causes an extra deployment change though.  Is this worth doing?

**Review by chrisohaver — APPROVED** [review]
LGTM

**Comment by rosti** [comment]
My point here is that during upgrades we should keep the behavior of CoreDNS the same as in the previous Kubernetes/CoreDNS version. Migration should only result in a new config if there are some unavoidable changes - config format change, drop of unsupported plugin and its replacement with a supported one.
In short a failed migration is a genuine error as I expect that the old config won't work with the new CoreDNS. Doing unnecessary migrations during upgrade (including replacing one default config with another) has the ability to change the CoreDNS behavior in between versions. Thus a cluster might get broken, slower, etc.
@neolit123 @fabriziopandini @ereslibre WDYT?

**Comment by neolit123** [comment]
it makes sense to me to only migrate when actually needed, yet it don't know how easy it is to distinguish that.

**Comment by chrisohaver** [comment]
> In short a failed migration is a genuine error as I expect that the old config won't work with the new CoreDNS.

@rosti, In this PR, if  `migration.Migrate()` returns an error, we leave CoreDNS unchanged: at the prior version, running on the old config.  You are kind of implying that is not the case, so just making sure this is understood.  Prior to this PR, this was not the case due to unintended code bug, not intentional design.

`migration.Migrate()`, already does very close to what you describe.  As called, it makes the following changes to a Corefile:
1. It removes plugins and options that are no longer recognized and if left in place would cause the new version of CoreDNS to fail to start.  These are absolutely required.
2. It removes options from plugins that are ignored by CoreDNS.  This could be skipped in theory.   But removing an option that is ignored will not change behavior.
3. It removes comments and standardizes indentation. This behavior is due to the way the corefile is parsed, and is hard to change. However, changes to indentation and comments will not functionally change a Corefile.

2 and 3 are IMO minor points because they do not change functionality. However, we could avoid them in part by inspecting the list of deprecations first. e.g. We could make the call to `migration.Deprecated()` earlier, loop through the results and only proceed to attempt to migrate if we see any "removed" options/plugins. However _I don't think this applies to the scope of this PR_, which is a bug fix. 

This PR aims to fix a bug in the current released version of kubeadm that had 2 consequences:
1.  After a successful corefile migration, the current deployment of CoreDNS was not pointed to the backup copy, such that if the Deployment rollout of the new version of CoreDNS failed, a rollback could fail (if the new config was not compatible with the old version).
2. After a failed corefile migration, the deployment of CoreDNS was left in a failed rollout state. This was issue #88725.

**Comment by rosti** [comment]
Thanks for the clarification @chrisohaver !

> > In short a failed migration is a genuine error as I expect that the old config won't work with the new CoreDNS.
> 
> @rosti, In this PR, if `migration.Migrate()` returns an error, we leave CoreDNS unchanged: at the prior version, running on the old config. You are kind of implying that is not the case, so just making sure this is understood. Prior to this PR, this was not the case due to unintended code bug, not intentional design.
> 

We seem to be on slightly different pages here. I was shocked that a fix for a failed migration is to leave the old config intact and the new CoreDNS version would simply pick it up and use it. The question I was asking myself is why did we need to attempt to perform the migration in first place? However, I saw that you already answered that below.

> Regarding only making changes when necessary: `migration.Migrate()`, already does very close to what you describe. As called, it makes the following changes to a Corefile:
> 
> 1. It removes plugins and options that are no longer recognized and if left in place would cause the new version of CoreDNS to fail to start.  These are absolutely required.
> 2. It removes options from plugins that are ignored by CoreDNS.  This could be skipped in theory.   But removing an option that is ignored will not change behavior.
> 3. It removes comments and standardizes indentation. This behavior is due to the way the corefile is parsed, and is hard to change. However, changes to indentation and comments will not functionally change a Corefile.
> 

The interesting case here is 1. It's obvious - CoreDNS won't start if we have an old plugin.
2 helps us maintain good Corefile hygiene and 3 might be considered a bit dangerous (upon upgrade we remove user's comments).

> 2 and 3 are IMO minor points because they do not change functionality. However, we could avoid them in part by inspecting the list of deprecations first. e.g. We could make the call to `migration.Deprecated()` earlier, loop through the results and only proceed to attempt to migrate if we see any "removed" options/plugins. However _I don't think this applies to the scope of this PR_, which is a bug fix.
> 

That's ideal for us. We don't want to silently skip telling the user that a migration error happened when the migration was supposed to remove an old plugin and make the config working with a the new CoreDNS version.

@rajansandeep Now that we are on the same page, I think the only thing left to do here is to make the earlier `migration.Deprecated()` call with a loop (as @chrisohaver suggested).

**Review by rosti — APPROVED** [review]
Thanks @rajansandeep !
I think, that I am happy with it. Can you please, squash your commits, so I can apply the LGTM label?

**Review by rosti — APPROVED** [review]
Thanks @rajansandeep !
/lgtm

**Pull request merged**